Fixes download free RECAP email documents using magic links

albertisfu opened the pull request:
- Added support to the `download_pdf_magic_link` new FreeOpinionReport from Juriscraper that allows us to download free  RECAP email documents using magic links without being logged in to PACER.

- If a RECAP Email contains a document URL without a magic number we avoid downloading the document from PACER.
- Added new magic_number argument to `fetch_pacer_doc_by_rd` and `download_pacer_pdf_by_rd` task in order to support new `download_pdf_magic_link` method, this new argument is option and by default is None in order to don't affect other functions that calls these tasks.
- Added a test to check new method download_pdf_magic_link is called when we try to fetch a RECAP email document with a magic number and added other a test to check we don't try to download the RECAP email document if the magic link is invalid or expired, It seems tests didn't pass, I think it's because I'm using a couple of email examples stored in S3 for these tests, maybe key access issue, I'm wondering if I have to change these tests to mock the reading S3 email file function instead of getting them directly from S3? 
- This is based on some Juriscraper changes https://github.com/freelawproject/juriscraper/pull/506 so for now, I've to install Juriscraper from the GitHub branch, I think it would be necessary to update Juriscraper package and then I can add the updated Juriscraper version in CL

mlissner commented on cl/recap/tasks.py:
```diff
@@ -1538,11 +1544,15 @@ def process_recap_email(
-        fetch_pacer_doc_by_rd(rd.pk, fq.pk)
+
+        # If we don't have a magic number avoid fetching the document
+        if magic_number:
+            fetch_pacer_doc_by_rd(rd.pk, fq.pk, magic_number)
```
Later we'll want to send an email here, to tell the user, "Hey, you got a notification, but it didn't have a magic link, so we weren't able to get the PDF. Still, here's the contents of your alert."

albertisfu replied:
Sure, I added a TODO comment here and I'll open an issue to work on it later.

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -1504,15 +1505,27 @@ def download_pacer_pdf_by_rd(
     report = FreeOpinionReport(pacer_court_id, s)
     try:
-        r = report.download_pdf(pacer_case_id, pacer_doc_id)
+        if magic_number:
+            # If a magic_number is passed as an argument try to download the
+            # document anonymously by magic link
+            de_seq_num = rd.docket_entry.pacer_sequence_number
+            r = report.download_pdf_magic_link(
+                pacer_case_id, pacer_doc_id, de_seq_num, magic_number
+            )
+        else:
+            # If no magic_number passed use normal method to fetch the document
+            r = report.download_pdf(pacer_case_id, pacer_doc_id)
```
I'm surprised here that you can't just use something like:

    r = report.download_pdf(pacer_case_id, pacer-doc_id, magic_number)

The one missing piece is the de_seq_num, but is that necessary? If so, could we add it to the `download_pdf` method more easily than adding a whole other method for it in juriscraper?

albertisfu replied:
Changed!

mlissner reviewed (changes requested):
Thanks, Alberto, this is great. Two things. One, I have a major design question about whether Juriscraper needs a whole other method (see below). 

Two, if it does (or if Juriscraper just needs tweaks), let's get that PR in first, release a new version of Juriscraper, and then update poetry with that. 

Other than that, just a few comments. I fear the one about factories will take a moment to get right, but I think the others are pretty small ones.

albertisfu commented on cl/recap/tests.py:
```diff
@@ -470,24 +471,45 @@ def test_pq_filters(self) -> None:
-        with open(os.path.join(test_dir, "recap_mail_receipt.json")) as file:
+        test_dir = Path(settings.INSTALL_ROOT) / "cl" / "recap" / "test_assets"
+        with open(
+            Path.joinpath(test_dir, "recap_mail_receipt.json"),
```
changed!

albertisfu commented on cl/recap/tests.py:
```diff
@@ -470,24 +471,45 @@ def test_pq_filters(self) -> None:
+            Path.joinpath(test_dir, "recap_mail_receipt.json"),
+            encoding="utf-8",
+        ) as file:
             recap_mail_receipt = json.loads(file.read())
```
changed!

albertisfu commented on cl/recap/tests.py:
```diff
@@ -470,24 +471,45 @@ def test_pq_filters(self) -> None:
 class RecapEmailToEmailProcessingQueueTest(TestCase):
     """Test the rest endpoint, but exclude the processing tasks."""
 
-    fixtures = ["canb_court.json"]
```
I swapped this fixture to use a CourtFactory, seems it's working well.

mlissner replied:
Can you remove the fixture too?

albertisfu commented:
These changes are ready, now we are using the same `download_pdf` method, I modified the method on Juriscraper https://github.com/freelawproject/juriscraper/pull/506 and deleted the other one.

- I deleted the tests I had because those only were checking if the other function was called if we had a magic_number, now we are using the same function, however, I'll be back here later to work on mocking S3 uploads and downloads project and design a new test for this.

- I'll open an issue about the function to send an email to the user if we receive an email without a magic link, to work on it later.

- I think I found a bug related to sending the docket to the webhook recipients, I'll review it in detail and open an issue too if it's necessary and solve it.

mlissner commented on cl/recap/tests.py:
```diff
@@ -522,6 +527,7 @@ def test_email_processing_queue_create(self):
         self.assertEqual(EmailProcessingQueue.objects.count(), 0)
         self.client.post(self.path, self.data, format="json")
         self.assertEqual(EmailProcessingQueue.objects.count(), 1)
+        # self.assertEqual(0, 1)
```
Stray debug line.

mlissner reviewed (changes requested):
Just a couple more very small, easy things. Can you please update the pyproject and poetry.lock files to use the next version of Juriscraper as well? Then, once it's released, we can merge this without worry?

albertisfu commented:
I think this is ready,  I've updated Juriscraper to the new 2.4.11 version.
- Removed the canb_court.json and fjc_data.json (this one was linked with canb_court.json) and replaced them with Factories.
- Rebased with main

albertisfu commented:
This seems ready, I've rebased with main and updated `Juriscraper` to 2.4.11.

Replaced `canb_court.json` with a factory and also was necessary to replace `fjc_data.json` because it was in the same test `IdbMergeTest`, for these factories I just added the required values to pass the test and avoid having a large factory.

mlissner commented:
Tests aren't happy. :(

albertisfu commented:
sorry, this was a weird one, I wasn’t getting this tests error in my dev machine, since it seems it has something to do with tests running in parallel, I only could reproduce the error by reducing the number of cores where docker is running on, apparently with more cores the probability the error occurs decrease.

The explanation is the following in case is useful: 
I was creating a court fixture from a CourtFactory for `cl.recap.tests.py-IdbMergeTest`
`cls.court = CourtFactory(id="canb", jurisdiction="FB") `

While in :
`cl.tests.test_factories.TestFactoryCreation`

For testing the creation of a [FjcIntegratedDatabaseFactory](https://github.com/freelawproject/courtlistener/blob/aafc3d5f3b1a6f021d93633f6d6a4a7f2c562d93/cl/recap/factories.py#L10)

It uses an Iterator to assign the `Court`:

```
circuit = Iterator(Court.objects.all())
district = Iterator(Court.objects.all())
```
Apparently, the list of `Courts` it was iterating included the one that was previously created on `IdbMergeTest` from the `CourtFactory`: `canb`

Seems once` IdbMergeTest` finished, the DB rolled back and `canb` court was not present in DB anymore, then when tried to finish the object creation of `FjcIntegratedDatabaseFactory`  it failed because the `canb` Court returned by the Iterator didn’t exist anymore in DB.

So to solve this, instead of creating a new Court with `CourtFactory` I just return an existing Court from the initial court data and use it to assign it to the `DocketFactory` and `FjcIntegratedDatabaseFactory` in this test.

`cls.court = Court.objects.get(id="scotus")`

mlissner commented:
Merged, thank you. Once this is deployed, we re-enable the lambda, and we're back in action?

albertisfu commented:
Sure, let me know once this is deployed and I'll reactivate the lambda function.
